draft-irtf-cfrg-fiat-shamir.md

```diff
@@ -107,59 +113,128 @@ This will be expanded in future versions of this specification.
+
+        def prove_batchable(self, witness, rng):
+            # Proving method using commitment-response format.
+            # Allows for batching.
+            (commitment, challenge, response) = self._prove(witness, rng)
+            # running the verifier here is just a sanity check
+            assert self.sigma_protocol.verifier(commitment, challenge, response)
+            domain_sep = "NISigmaProtocolBatchableProof".encode('utf-8')
+            return domain_sep + self.sigma_protocol.serialize_commitment(commitment) + self.sigma_protocol.serialize_response(response)
+
+        def verify_batchable(self, proof):
+            # Before running the sigma protocol verifier, one must also check that:
+            # - serialized proof domain separator matches expected value
+            domain_sep = "NISigmaProtocolBatchableProof".encode('utf-8')
+            domain_sep_bytes = proof[:len(domain_sep)]
+            assert domain_sep_bytes == domain_sep
+
+            # - the proof length is exactly len(domain_sep) + commit_bytes_len + response_bytes_len
+            assert len(proof) == len(domain_sep) + self.sigma_protocol.instance.commit_bytes_len + self.sigma_protocol.instance.response_bytes_len
 
-## Codec for Linear maps {#group-prove}
+            # - proof deserialization successfully produces a valid commitment and a valid response
+            commitment_bytes = proof[len(domain_sep):self.sigma_protocol.instance.commit_bytes_len]
+            response_bytes = proof[len(domain_sep) + self.sigma_protocol.instance.commit_bytes_len:]
+            commitment = self.sigma_protocol.deserialize_commitment(commitment_bytes)
+            response = self.sigma_protocol.deserialize_response(response_bytes)
 
-We describe a codec for Schnorr proofs over groups of prime order `p` that is intended for duplex sponges where `Unit = u8`.
+            self.codec.prover_message(self.hash_state, commitment)
+            challenge = self.codec.verifier_challenge(self.hash_state)
+            return self.sigma_protocol.verifier(commitment, challenge, response)
 
-    class LinearMapCodec:
-        Group: groups.Group = None
-        DuplexSponge: DuplexSpongeInterface = None
+## NIZK instances (ciphersuites)
 
-        def init(self, iv: bytes):
-            self.hash_state = self.DuplexSponge(iv)
+We describe NIZK instances for combinations of protocols (SigmaProtocol), codec (Codec), and hash fuction (DuplexSpongeInterface). Descriptions of codecs and hash functions are in the following sections.
+
+    class NISchnorrProofShake128P256(NIZK):
+        Protocol = SchnorrProof
+        Codec = P256Codec
+        Hash = SHAKE128
+
+    class NISchnorrProofShake128Bls12381(NIZK):
+        Protocol = SchnorrProof
+        Codec = Bls12381Codec
+        Hash = SHAKE128
+
+    class NISchnorrProofKeccakDuplexSpongeBls12381(NIZK):
+        Protocol = SchnorrProof
+        Codec = Bls12381Codec
+        Hash = KeccakDuplexSponge
```

Comment:
I'm curious why we're defining this variant. What's the advantage of duplex over plain old SHAKE128?

Reply:
@mmaker could you answer this one? I am just copying this over from the poc since it seemed like they had gotten out of date :)
I'm okay skipping defining this in the spec if it seems unnecessary though, I'm all for reducing complexity :)

Reply:
It makes sense to keep if someone is using it.

Reply:
@cjpatton thanks for bringing it up! let's discuss this in a separate issue, I agree with you but wanted to share more state about this